src/elfpy/user.py

```diff
@@ -23,54 +30,92 @@ class User:
-        self.wallet = self.UserWallet(base_in_wallet=self.budget)
+        self.wallet = AgentWallet(base_in_wallet=self.budget)
         self.rng = rng
+        self.wallet_address = wallet_address
         self.verbose = verbose
         self.last_update_spend = 0
+        self.product_of_time_and_base = 0
         self.weighted_average_spend = 0
-
-    @dataclass
-    class UserWallet:
-        """user wallet store"""
-
-        # pylint: disable=missing-function-docstring
-        base_in_wallet: float = 0
-        token_in_wallet: dict = field(default_factory=dict)
-        base_in_protocol: dict = field(default_factory=dict)
-        token_in_protocol: dict = field(default_factory=dict)
-
-        def __getitem__(self, key):
-            return getattr(self, key)
-
-        def __setitem__(self, key, value):
-            setattr(self, key, value)
-
-        def update(self, *args, **kwargs):
-            return self.__dict__.update(*args, **kwargs)
-
-        def keys(self):
-            return self.__dict__.keys()
-
-        def values(self):
-            return self.__dict__.values()
-
-        def items(self):
-            return self.__dict__.items()
+        self.position_list = []
+        self.status_update()
 
     @dataclass
     class UserAction:
         """user action specification"""
 
         action_type: str
         trade_amount: float
-        mint_time: float = field(default=None)
+        wallet_address: int
+        agent: object
+        mint_time: float = None
+        fee_percent: float = field(init=False)
+        init_share_price: float = field(init=False)
+        share_price: float = field(init=False)  
+        share_reserves: float = field(init=False)
+        bond_reserves: float = field(init=False)
+        share_buffer: float = field(init=False)
+        bond_buffer: float = field(init=False)
+        liquidity_pool: float = field(init=False)
+        rate: float = field(init=False)
+        time_remaining: float = field(init=False)
+        stretched_time_remaining: float = field(init=False)
+
+        def print_description_string(self):
+            output_string = f"🤖 {bcolors.FAIL}{self.wallet_address}{bcolors.ENDC}"
+            # for key, value in self.__dataclass_fields__.items():
+            for key, value in self.__dict__.items():
+                if key == "action_type":
+                    output_string += f" execute {bcolors.FAIL}{value}(){bcolors.ENDC}"
+                elif key not in ["wallet_address","agent"]:
+                    output_string += f" {key}: "
+                    if value<2:
+                        output_string += f"{value:.5f}"
+                    elif value<100:
+                        output_string += f"{value:.2f}"
+                    else:
+                        output_string += f"{value:,.0f}"
+            print(output_string)
+
+        def update_market_variables(self, market, wallet_address=None):
```

Comment:
Is there a better way to do this? The user has access to self.market. So as long as you're careful about passing `user.market` as an arg, you should automatically have access to the latest market variables. I don't like that we now have a bunch of user member variables that are manually set to track market member variables.

Reply:
9 were passthrough assignments, moved those to where they're called in the _long/_short functions. 3 pieces of logic i moved out to where the function was.